diff --git a/configure b/configure
index 99dd1fe..01051ea 100755
--- a/configure
+++ b/configure
@@ -6948,6 +6948,8 @@ rm -f conftest.c conftest.out
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for multiarch" >&5
 printf %s "checking for multiarch... " >&6; }
 case $ac_sys_system in #(
+  Linux-android) :
+    MULTIARCH="" ;; #(
   Darwin*) :
     MULTIARCH="" ;; #(
   FreeBSD*) :
@@ -12852,7 +12854,7 @@ then
 		 then CCSHARED="-fPIC";
 		 else CCSHARED="+z";
 		 fi;;
-	Linux-android*) ;;
+	Linux-android*) CCSHARED="-fPIC";;
 	Linux*|GNU*) CCSHARED="-fPIC";;
 	Emscripten*|WASI*)
 		if test "x$enable_wasm_dynamic_linking" = xyes
@@ -18786,7 +18788,6 @@ printf "%s\n" "$ac_cv_func_link" >&6; }
   if test "x$ac_cv_func_link" = xyes
 then :
 
-printf "%s\n" "#define HAVE_LINK 1" >>confdefs.h
 
 fi
 
