diff --git a/Lib/mailcap.py b/Lib/mailcap.py
index 2f4656e..0db23c2 100644
--- a/Lib/mailcap.py
+++ b/Lib/mailcap.py
@@ -67,7 +67,8 @@ def listmailcapfiles():
             # Don't bother with getpwuid()
             home = '.' # Last resort
         mailcaps = [home + '/.mailcap', '/etc/mailcap',
-                '/usr/etc/mailcap', '/usr/local/etc/mailcap']
+                '/usr/etc/mailcap', '/usr/local/etc/mailcap',
+                '@TERMUX_PREFIX@/etc/mailcap']
     return mailcaps
 
 
