diff --git a/Lib/mimetypes.py b/Lib/mimetypes.py
index 37228de..6bdf5dc 100644
--- a/Lib/mimetypes.py
+++ b/Lib/mimetypes.py
@@ -55,6 +55,7 @@ knownfiles = [
     "/usr/local/lib/netscape/mime.types",
     "/usr/local/etc/httpd/conf/mime.types",     # Apache 1.2
     "/usr/local/etc/mime.types",                # Apache 1.3
+    "@TERMUX_PREFIX@/etc/mime.types",           # Termux
     ]
 
 inited = False
