diff --git a/Lib/http/server.py b/Lib/http/server.py
index ca6240d..b4f3b97 100644
--- a/Lib/http/server.py
+++ b/Lib/http/server.py
@@ -1183,10 +1183,6 @@ class CGIHTTPRequestHandler(SimpleHTTPRequestHandler):
                 return
             # Child
             try:
-                try:
-                    os.setuid(nobody)
-                except OSError:
-                    pass
                 os.dup2(self.rfile.fileno(), 0)
                 os.dup2(self.wfile.fileno(), 1)
                 os.execve(scriptfile, args, env)
