diff --git a/Lib/subprocess.py b/Lib/subprocess.py
index 6df5dd5..d7ef1e2 100644
--- a/Lib/subprocess.py
+++ b/Lib/subprocess.py
@@ -1808,9 +1808,7 @@ class Popen:
                 args = list(args)
 
             if shell:
-                # On Android the default shell is at '/system/bin/sh'.
-                unix_shell = ('/system/bin/sh' if
-                          hasattr(sys, 'getandroidapilevel') else '/bin/sh')
+                unix_shell = ('@TERMUX_PREFIX@/bin/sh')
                 args = [unix_shell, "-c"] + args
                 if executable:
                     args[0] = executable
