diff --git a/Lib/aifc.py b/Lib/aifc.py
index 5254987..226fe59 100644
--- a/Lib/aifc.py
+++ b/Lib/aifc.py
@@ -961,7 +961,7 @@ def open(f, mode=None):
 if __name__ == '__main__':
     import sys
     if not sys.argv[1:]:
-        sys.argv.append('/usr/demos/data/audio/bach.aiff')
+        sys.argv.append('@TERMUX_PREFIX@/demos/data/audio/bach.aiff')
     fn = sys.argv[1]
     with open(fn, 'r') as f:
         print("Reading", fn)
