--- a/Lib/mimetypes.py
+++ b/Lib/mimetypes.py
@@ -49,6 +49,7 @@
     "/usr/local/lib/netscape/mime.types",
     "/usr/local/etc/httpd/conf/mime.types",     # Apache 1.2
     "/usr/local/etc/mime.types",                # Apache 1.3
+    "/data/data/com.termux/files/usr/etc/mime.types",           # Termux
     ]
 
 inited = False
--- a/Lib/posixpath.py
+++ b/Lib/posixpath.py
@@ -32,7 +32,7 @@
 extsep = '.'
 sep = '/'
 pathsep = ':'
-defpath = '/bin:/usr/bin'
+defpath = '/data/data/com.termux/files/usr/bin'
 altsep = None
 devnull = '/dev/null'
 
--- a/Lib/uuid.py
+++ b/Lib/uuid.py
@@ -361,7 +361,6 @@

     try:
         path_dirs = os.environ.get('PATH', os.defpath).split(os.pathsep)
-        path_dirs.extend(['/sbin', '/usr/sbin'])
         executable = shutil.which(command, path=os.pathsep.join(path_dirs))
         if executable is None:
             return None
--- a/Lib/subprocess.py
+++ b/Lib/subprocess.py
@@ -1389,9 +1389,7 @@
                 args = list(args)
 
             if shell:
-                # On Android the default shell is at '/system/bin/sh'.
-                unix_shell = ('/system/bin/sh' if
-                          hasattr(sys, 'getandroidapilevel') else '/bin/sh')
+                unix_shell = ('/data/data/com.termux/files/usr/bin/sh')
                 args = [unix_shell, "-c"] + args
                 if executable:
                     args[0] = executable
--- a/Lib/tempfile.py
+++ b/Lib/tempfile.py
@@ -170,7 +170,7 @@ def _candidate_tempdir_list():
                          _os.path.expandvars(r'%SYSTEMROOT%\Temp'),
                          r'c:\temp', r'c:\tmp', r'\temp', r'\tmp' ])
     else:
-        dirlist.extend([ '/tmp', '/var/tmp', '/usr/tmp' ])
+        dirlist.extend([ '/data/data/com.termux/files/usr/tmp' ])
 
     # As a last resort, the current directory.
     try:
--- a/Lib/http/server.py
+++ b/Lib/http/server.py
@@ -1165,10 +1165,6 @@
                 return
             # Child
             try:
-                try:
-                    os.setuid(nobody)
-                except OSError:
-                    pass
                 os.dup2(self.rfile.fileno(), 0)
                 os.dup2(self.wfile.fileno(), 1)
                 os.execve(scriptfile, args, env)
--- a/Lib/ctypes/util.py
+++ b/Lib/ctypes/util.py
@@ -172,7 +172,7 @@
             # assuming GNU binutils / ELF
             if not f:
                 return None
-            objdump = shutil.which('objdump')
+            objdump = shutil.which('llvm-objdump')
             if not objdump:
                 # objdump is not available, give up
                 return None
@@ -209,7 +209,7 @@
             expr = os.fsencode(expr)
 
             try:
-                proc = subprocess.Popen(('/sbin/ldconfig', '-r'),
+                proc = subprocess.Popen(('/data/data/com.termux/files/usr/bin/ldconfig', '-r'),
                                         stdout=subprocess.PIPE,
                                         stderr=subprocess.DEVNULL)
             except OSError:  # E.g. command not found
@@ -300,7 +300,7 @@
         def _findLib_ld(name):
             # See issue #9998 for why this is needed
             expr = r'[^\(\)\s]*lib%s\.[^\(\)\s]*' % re.escape(name)
-            cmd = ['ld', '-t']
+            cmd = ['ld.lld', '-t']
             libpath = os.environ.get('LD_LIBRARY_PATH')
             if libpath:
                 for d in libpath.split(':'):
--- a/Modules/socketmodule.c
+++ b/Modules/socketmodule.c
@@ -5373,17 +5373,74 @@
    This only returns the protocol number, since the other info is
    already known or not useful (like the list of aliases). */
 
+#ifdef __ANDROID__
+struct protocol_name_and_number {
+    char* name;
+    int number;
+};
+#endif
+
 /*ARGSUSED*/
 static PyObject *
 socket_getprotobyname(PyObject *self, PyObject *args)
 {
+#ifdef __ANDROID__
+    /* http://git.musl-libc.org/cgit/musl/tree/src/network/proto.c */
+    static const struct protocol_name_and_number protocols[] = {
+       {"ip", 0},
+       {"icmp", 1},
+       {"igmp", 2},
+       {"ggp", 3},
+       {"ipencap", 4},
+       {"st", 5},
+       {"tcp", 6},
+       {"egp", 8},
+       {"pup", 12},
+       {"udp", 17},
+       {"hmp", 20},
+       {"xns-idp", 22},
+       {"iso-tp4", 29},
+       {"xtp", 36},
+       {"ddp", 37},
+       {"idpr-cmtp", 38},
+       {"ipv6", 41},
+       {"ipv6-route", 43},
+       {"ipv6-frag", 44},
+       {"idrp", 45},
+       {"rsvp", 46},
+       {"gre", 47},
+       {"esp", 50},
+       {"ah", 51},
+       {"skip", 57},
+       {"ipv6-icmp", 58},
+       {"ipv6-nonxt", 59},
+       {"ipv6-opts", 60},
+       {"rspf", 73},
+       {"vmtp", 81},
+       {"ospf", 89},
+       {"ipip", 94},
+       {"encap", 98},
+       {"pim", 103},
+       {"raw", 255}
+    };
+    int i;
+#endif
     const char *name;
     struct protoent *sp;
     if (!PyArg_ParseTuple(args, "s:getprotobyname", &name))
         return NULL;
+#ifdef __ANDROID__
+    for (i = 0; i < sizeof(protocols) / sizeof(protocols[0]); i++) {
+        if (strcmp(protocols[i].name, name) == 0) {
+            return PyLong_FromLong((long) protocols[i].number);
+        }
+    }
+    sp = NULL;
+#else
     Py_BEGIN_ALLOW_THREADS
     sp = getprotobyname(name);
     Py_END_ALLOW_THREADS
+#endif
     if (sp == NULL) {
         PyErr_SetString(PyExc_OSError, "protocol not found");
         return NULL;
--- a/Lib/multiprocessing/heap.py
+++ b/Lib/multiprocessing/heap.py
@@ -70,7 +70,7 @@
         """
 
         if sys.platform == 'linux':
-            _dir_candidates = ['/dev/shm']
+            _dir_candidates = []
         else:
             _dir_candidates = []
 

--- a/Modules/_multiprocessing/posixshmem.c
+++ b/Modules/_multiprocessing/posixshmem.c
@@ -16,6 +16,71 @@ posixshmem - A Python extension that provides shm_open() and shm_unlink()
 #  include <sys/mman.h>           // shm_open(), shm_unlink()
 #endif
 
+#ifdef __ANDROID__
+#include <alloca.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+static int shm_unlink(const char *name) {
+    size_t namelen;
+    char *fname;
+
+    /* Construct the filename.  */
+    while (name[0] == '/') ++name;
+
+    if (name[0] == '\0') {
+        /* The name "/" is not supported.  */
+        errno = EINVAL;
+        return -1;
+    }
+
+    namelen = strlen(name);
+    fname = (char *) alloca(sizeof("/data/data/com.termux/files/usr/tmp/") - 1 + namelen + 1);
+    memcpy(fname, "/data/data/com.termux/files/usr/tmp/", sizeof("/data/data/com.termux/files/usr/tmp/") - 1);
+    memcpy(fname + sizeof("/data/data/com.termux/files/usr/tmp/") - 1, name, namelen + 1);
+
+    return unlink(fname);
+}
+
+static int shm_open(const char *name, int oflag, mode_t mode) {
+    size_t namelen;
+    char *fname;
+    int fd;
+
+    /* Construct the filename.  */
+    while (name[0] == '/') ++name;
+
+    if (name[0] == '\0') {
+        /* The name "/" is not supported.  */
+        errno = EINVAL;
+        return -1;
+    }
+
+    namelen = strlen(name);
+    fname = (char *) alloca(sizeof("/data/data/com.termux/files/usr/tmp/") - 1 + namelen + 1);
+    memcpy(fname, "/data/data/com.termux/files/usr/tmp/", sizeof("/data/data/com.termux/files/usr/tmp/") - 1);
+    memcpy(fname + sizeof("/data/data/com.termux/files/usr/tmp/") - 1, name, namelen + 1);
+
+    fd = open(fname, oflag, mode);
+    if (fd != -1) {
+        /* We got a descriptor.  Now set the FD_CLOEXEC bit.  */
+        int flags = fcntl(fd, F_GETFD, 0);
+        flags |= FD_CLOEXEC;
+        flags = fcntl(fd, F_SETFD, flags);
+
+        if (flags == -1) {
+            /* Something went wrong.  We cannot return the descriptor.  */
+            int save_errno = errno;
+            close(fd);
+            fd = -1;
+            errno = save_errno;
+        }
+    }
+
+    return fd;
+}
+#endif
 
 /*[clinic input]
 module _posixshmem
--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -310,7 +310,7 @@ corresponding Unix manual entries for more information on calls.");
 #endif
 
 #if defined(HAVE_SYS_XATTR_H)
-#  if defined(HAVE_LINUX_LIMITS_H) && !defined(__FreeBSD_kernel__) && !defined(__GNU__)
+#  if defined(HAVE_LINUX_LIMITS_H) && !defined(__FreeBSD_kernel__) && !defined(__GNU__) && !defined(__ANDROID__)
 #    define USE_XATTRS
 #    include <linux/limits.h>  // Needed for XATTR_SIZE_MAX on musl libc.
 #  endif
--- a/configure
+++ b/configure
@@ -7002,6 +7002,8 @@ rm -f conftest.out
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for multiarch" >&5
 printf %s "checking for multiarch... " >&6; }
 case $ac_sys_system in #(
+  Linux-android) :
+    MULTIARCH="" ;; #(
   Darwin*) :
     MULTIARCH="" ;; #(
   iOS) :
--- a/configure
+++ b/configure
@@ -18772,7 +18774,6 @@
   if test "x$ac_cv_func_link" = xyes
 then :
 
-printf "%s\n" "#define HAVE_LINK 1" >>confdefs.h
 
 fi
 
--- a/configure
+++ b/configure
@@ -13048,6 +13048,7 @@ then
 		 then CCSHARED="-fPIC";
 		 else CCSHARED="+z";
 		 fi;;
+	Linux-android*) CCSHARED="-fPIC";;
 	Linux*|GNU*) CCSHARED="-fPIC";;
 	Emscripten*|WASI*)
 		if test "x$enable_wasm_dynamic_linking" = xyes
